DeviceAdapters/Zaber/WdiAutofocus.cpp

```diff
@@ -0,0 +1,340 @@
+{
+	this->LogMessage("WdiAutofocus::FocusAddressGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		if (initialized_)
+		{
+			resetConnection();
+		}
+
+		pProp->Get(focusAddress_);
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		pProp->Set(focusAddress_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::FocusAxisGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	this->LogMessage("WdiAutofocus::FocusAxisGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		if (initialized_)
+		{
+			resetConnection();
+		}
+
+		pProp->Get(focusAxis_);
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		pProp->Set(focusAxis_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::LimitMinGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	this->LogMessage("ObjectiveChanger::LimitMinGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		double newLimit;
+		pProp->Get(newLimit);
+		auto update = limitMin_ != newLimit;
+		limitMin_ = newLimit;
+		if (update) {
+			return handleException([=]() {
+				ensureConnected();
+				axis_.getSettings().set("motion.tracking.limit.min", limitMin_ * WdiAutofocus_::xLdaNativePerMm);
+			});
+		}
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		if (initialized_)
+		{
+			int ret = handleException([=]() {
+				ensureConnected();
+				limitMin_ = axis_.getSettings().get("motion.tracking.limit.min") / WdiAutofocus_::xLdaNativePerMm;
+			});
+			if (ret != DEVICE_OK)
+			{
+				return ret;
+			}
+		}
+
+		pProp->Set(limitMin_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::LimitMaxGetSet(MM::PropertyBase* pProp, MM::ActionType eAct)
+{
+	this->LogMessage("ObjectiveChanger::LimitMaxGetSet\n", true);
+
+	if (eAct == MM::AfterSet)
+	{
+		double newLimit;
+		pProp->Get(newLimit);
+		auto update = limitMax_ != newLimit;
+		limitMax_ = newLimit;
+		if (update) {
+			return handleException([=]() {
+				ensureConnected();
+				axis_.getSettings().set("motion.tracking.limit.max", limitMax_ * WdiAutofocus_::xLdaNativePerMm);
+			});
+		}
+	}
+	else if (eAct == MM::BeforeGet)
+	{
+		if (initialized_)
+		{
+			int ret = handleException([=]() {
+				ensureConnected();
+				limitMax_ = axis_.getSettings().get("motion.tracking.limit.max") / WdiAutofocus_::xLdaNativePerMm;
+			});
+			if (ret != DEVICE_OK)
+			{
+				return ret;
+			}
+		}
+
+		pProp->Set(limitMax_);
+	}
+
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::FullFocus() {
+	this->LogMessage("WdiAutofocus::FullFocus\n", true);
+
+	return handleException([=]() {
+		ensureConnected();
+
+		axis_.genericCommand("move track once");
+		axis_.waitUntilIdle();
+	});
+}
+
+int WdiAutofocus::IncrementalFocus() {
+	this->LogMessage("WdiAutofocus::IncrementalFocus\n", true);
+
+	return FullFocus();
+}
+
+int WdiAutofocus::GetLastFocusScore(double& score) {
+	this->LogMessage("WdiAutofocus::GetLastFocusScore\n", true);
+	score = 0.0;
+	return DEVICE_OK;
+}
+
+int WdiAutofocus::GetCurrentFocusScore(double& score) {
+	this->LogMessage("WdiAutofocus::GetCurrentFocusScore\n", true);
+	score = 0.0;
```

Comment:
:(  I assume this defeats MM's software autofocus?

Reply:
There is very little documentation around the interface. I mostly inspired by the others. It's unclear where this gets actually used.